cpu/nrf51/periph/pwm.c

```diff
@@ -0,0 +1,224 @@
+    NRF_PPI->CHENSET = NRF51_PWM_PPI_CHANNELS;
+
+    /* shortcut to reset Counter after CC[1] event */
+    PWM_TIMER->SHORTS = TIMER_SHORTS_COMPARE1_CLEAR_Msk;
+
+    PWM_TIMER->TASKS_START = 1;
+
+    DEBUG("Timer frequency is set to %ld\n",
+          apply_prescaler(PWM_TIMER->PRESCALER));
+
+    return (uint32_t)(apply_prescaler(PWM_TIMER->PRESCALER) / res);
+}
+
+void pwm_set(pwm_t dev, uint8_t channel, uint16_t value)
+{
+    assert((dev == 0) && (channel == 0));
+
+    /*
+     * make sure duty cycle is set at the beggining of each period
+     * ensure to stop the timer as soon as possible
+     */
+    PWM_TIMER->TASKS_STOP = 1;
+    PWM_TIMER->EVENTS_COMPARE[1] = 0;
+    PWM_TIMER->SHORTS = TIMER_SHORTS_COMPARE1_STOP_Msk;
+    PWM_TIMER->TASKS_START = 1;
+
+    /*
+     * waiting for the timer to stop
+     * This loop generates heavy load. This is not optimal therefore a local
+     * sleep function should be implemented.
+     */
+    while (PWM_TIMER->EVENTS_COMPARE[1] == 0) {};
+
+    /*
+     * checking pwm alignment first
+     * and guarding if duty cycle is 0% / 100%
+     */
+    if (NRF_GPIOTE->CONFIG[PWM_GPIOTE_CH] & GPIOTE_CONFIG_OUTINIT_Msk) {
+        if (value == 0) {
+            if (PWM_TIMER->CC[0] != 0) {
+                NRF_GPIOTE->TASKS_OUT[PWM_GPIOTE_CH] = 1;
+            }
+
+            NRF_PPI->CHENCLR = NRF51_PWM_PPI_CHANNELS;
+            PWM_TIMER->CC[0] = 0;
+        } else {
+            if (PWM_TIMER->CC[0] == 0) {
+                NRF_GPIOTE->TASKS_OUT[PWM_GPIOTE_CH] = 1;
+            }
+            if (value >= PWM_TIMER->CC[1]) {
+                NRF_PPI->CHENCLR = NRF51_PWM_PPI_CHANNELS;
+                PWM_TIMER->CC[0] = PWM_TIMER->CC[1];
+            } else {
+                NRF_PPI->CHENSET = NRF51_PWM_PPI_CHANNELS;
+                PWM_TIMER->CC[0] = value;
+            }
+        }
+    } else {
+        if (value >= PWM_TIMER->CC[1]) {
+            if (PWM_TIMER->CC[0] != PWM_TIMER->CC[1]) {
+                NRF_GPIOTE->TASKS_OUT[PWM_GPIOTE_CH] = 1;
+            }
+            NRF_PPI->CHENCLR = NRF51_PWM_PPI_CHANNELS;
+            PWM_TIMER->CC[0] = PWM_TIMER->CC[1];
+        } else {
+            if (PWM_TIMER->CC[0] == PWM_TIMER->CC[1]) {
+                NRF_GPIOTE->TASKS_OUT[PWM_GPIOTE_CH] = 1;
+            }
+            if (value == 0) {
+                NRF_PPI->CHENCLR = NRF51_PWM_PPI_CHANNELS;
+                PWM_TIMER->CC[0] = 0;
+            } else {
+                NRF_PPI->CHENSET = NRF51_PWM_PPI_CHANNELS;
+                PWM_TIMER->CC[0] = PWM_TIMER->CC[1] - value;
+            }
+        }
+    }
+
+    /* reconfigure pwm to standard mode */
+    PWM_TIMER->EVENTS_COMPARE[1] = 0;
+    PWM_TIMER->TASKS_CLEAR = 1;
+    PWM_TIMER->SHORTS = TIMER_SHORTS_COMPARE1_CLEAR_Msk;
+    PWM_TIMER->TASKS_START = 1;
+}
+
+uint8_t pwm_channels(pwm_t dev)
+{
+    assert(dev == 0);
+    return 1;
+}
+
+void pwm_poweron(pwm_t dev)
+{
+    assert(dev == 0);
+
+#if CPU_FAM_NRF51
```

Comment:
this PWM driver is written specifically for the `nrf51`, so no need for this ifdef...